File: Crio/crio_programming_interview_problems-master/CobbledStreets/CobbledStreets.cpp
#include <bits/stdc++.h>
using namespace std;

// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
int getParent(int x, vector<int> &parent)
{
    if (parent[x] == x)
        return x;

    return parent[x] = getParent(parent[x], parent);
}

void setUnion(int a, int b, vector<int> &parent, vector<int> &size)
{
    a = getParent(a, parent);
    b = getParent(b, parent);

    if (size[a] > size[b])
    {
        swap(a, b);
    }

    parent[b] = a;
    size[a] += size[b];
    return;
}

bool comp(const vector<int> &a,const vector<int> &b){
    return a[2]<b[2];
}
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

int cobbledStreets(int n, int p, vector<vector<int> > &edges)
{
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    sort(edges.begin(), edges.end(),comp);
    vector<int> size(n, 0);

    long long cost = 0;

    vector<int> parent(n + 1);
    for (int i = 1; i <= n; i++)
    {
        parent[i] = i;
    }

    for (auto edge : edges)
    {
        int p1 = getParent(edge[0], parent);
        int p2 = getParent(edge[1], parent);

        if (p1 == p2)
        {
            continue;
        }
        cost += edge[2];

        setUnion(p1, p2, parent, size);
    }

    return cost * p;
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}

int main()
{

    int t;
    cin >> t;

    while (t--)
    {
        int p, n, m;
        cin >> p >> n >> m;

        vector<vector<int> > edges;

        for (int i = 0; i < m; i++)
        {
            int u, v, w;
            cin >> u >> v >> w;
            edges.push_back({u,v,w});
        }

        int ans = cobbledStreets(n, p, edges);
        cout << ans << endl;
    }
}
 

File: Crio/crio_programming_interview_problems-master/CobbledStreets/CobbledStreets.py
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
from collections import defaultdict 

class Graph:
  
    def __init__(self,vertices,edges): 
        self.V= vertices 
        self.graph = edges
        self.parent = [i for i in range(vertices+1)]

    def addEdge(self,u,v,w): 
        self.graph.append([u,v,w]) 

    def find(self, i, parent): 
        while parent[i] != i: 
            parent[i] = parent[parent[i]]
            i = parent[i]
        return i

    def union(self, parent, x, y): 
        xroot = self.find(x , parent) 
        yroot = self.find(y , parent) 
        self.parent[xroot] = yroot
        
    def KruskalMST(self): 
        result = list()
        cost = 0 
        self.graph =  sorted(self.graph,key=lambda item: item[2])
        for edge in self.graph:
            p1 = self.find(edge[0],self.parent)
            p2 = self.find(edge[1],self.parent)
            if p1==p2:
                continue
            cost += edge[2] 
            self.union(self.parent, p1,p2)
        return cost
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def cobbledStreets(n, p ,edges):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    edges = sorted(edges)
    graph = Graph(n,edges)
    cost = graph.KruskalMST()
    return cost*p
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def main():
    t = int(input())
    for i in range(t):
        p,n,m = map(int,input().strip().split())
        edges = list()
        for _ in range(m):
            edge = list(map(int,input().split()))
            edges.append(edge)
        ans = cobbledStreets(n,p,edges)
        print(ans)


if __name__ == "__main__":
    main()
